src/main/java/org/embulk/input/marketo/delegate/MarketoBaseBulkExtractInputPlugin.java

```diff
@@ -0,0 +1,199 @@
+            currentTimestamp = Timestamp.ofEpochMilli(task.getLatestFetchTime().get());
+        }
+        TimestampParser timestampParser = new TimestampParser(MarketoUtils.ISO_8601_FORMAT, DateTimeZone.UTC);
+        try (LineDecoder lineDecoder = new LineDecoder(new InputStreamFileInput(task.getBufferAllocator(), inputStream), task)) {
+            CsvTokenizer csvTokenizer = new CsvTokenizer(lineDecoder, task);
+            if (!csvTokenizer.nextFile()) {
+                throw new DataException("Can't read extract input stream");
+            }
+            csvTokenizer.nextRecord();
+            List<String> headers = new ArrayList<>();
+            while (csvTokenizer.hasNextColumn()) {
+                headers.add(csvTokenizer.nextColumn());
+            }
+            while (csvTokenizer.nextRecord()) {
+                List<String> values = new ArrayList<>();
+                while (csvTokenizer.hasNextColumn()) {
+                    values.add(csvTokenizer.nextColumnOrNull());
+                }
+                if (imported >= PREVIEW_RECORD_LIMIT && Exec.isPreview()) {
+                    break;
+                }
+                final Map<String, String> kvMap = MarketoUtils.zip(headers, values);
+                ObjectNode objectNode = MarketoUtils.transformToObjectNode(kvMap, pageBuilder.getSchema());
+                if (!objectNode.has(incrementalColumn)) {
+                    throw new DataException("Extracted record doesn't have incremental column " + incrementalColumn);
+                }
+                JsonNode incrementalJsonNode = objectNode.get(incrementalColumn);
+                if (uidColumn != null) {
+                    JsonNode uidField = objectNode.get(uidColumn);
+                    String uid = uidField.asText();
+                    if (latestUids.contains(uid)) {
+                        //Duplicate value
+                        continue;
+                    }
+                }
+                String incrementalTimeStamp = incrementalJsonNode.asText();
+                Timestamp timestamp = timestampParser.parse(incrementalTimeStamp);
+                if (currentTimestamp == null) {
+                    currentTimestamp = timestamp;
+                }
+                else {
+                    int compareTo = currentTimestamp.compareTo(timestamp);
+                    if (compareTo < 0) {
+                        currentTimestamp = timestamp;
+                        //switch timestamp
+                        latestUids.clear();
+                    }
+                    else if (compareTo == 0) {
+                        //timestamp is equal
+                        if (uidColumn != null) {
+                            latestUids.add(uidColumn);
+                        }
+                    }
+                }
+                recordImporter.importRecord(new JacksonServiceRecord(objectNode), pageBuilder);
+                imported++;
+            }
+        }
```

Comment:
The logic of this method seems to be very complicated to follow. I suggest to refactor it into readable blocks/methods for easier to read the code.